File: clipsync-win/ClipSync.Tests/StreamPlannerTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using ClipSync.Net;
using Xunit;

namespace ClipSync.Tests;

public class StreamPlannerTests
{
    private static ClipFormat Fmt(string mime, int size) =>
        new(mime, (ulong)size, new byte[size], null);

    private static ClipboardItem Item(params ClipFormat[] formats) =>
        new(1, new byte[32], 0, formats.ToList(), null);

    private static ulong _next = 100;
    private static ulong NextId() => _next++;

    private const int KiB = 1024;
    private const int MiB = 1024 * 1024;

    [Fact]
    public void Small_item_passes_through_inline_with_no_streams()
    {
        var item = Item(Fmt("text/plain", 10), Fmt("text/html", 64 * KiB));
        var plan = StreamPlanner.Plan(item, peerStreams: true, NextId);

        Assert.NotNull(plan.WireItem);
        Assert.Same(item, plan.WireItem);          // untouched object
        Assert.Empty(plan.Streams);
        Assert.Empty(plan.Dropped);
    }

    [Fact]
    public void Large_format_is_replaced_by_stream_id_and_queued()
    {
        var text = Fmt("text/plain", 10);
        var img = Fmt("image/png", 64 * KiB + 1);
        var plan = StreamPlanner.Plan(Item(text, img), peerStreams: true, NextId);

        var wire = plan.WireItem!;
        Assert.Equal(2, wire.Formats.Count);
        Assert.NotNull(wire.Formats[0].Inline);
        Assert.Null(wire.Formats[0].StreamId);
        Assert.Null(wire.Formats[1].Inline);
        Assert.NotNull(wire.Formats[1].StreamId);
        Assert.Equal(img.Size, wire.Formats[1].Size);

        var s = Assert.Single(plan.Streams);
        Assert.Equal(wire.Formats[1].StreamId, s.StreamId);
        Assert.Same(img.Inline, s.Data);
        Assert.Empty(plan.Dropped);
    }

    [Fact]
    public void Distinct_streams_get_distinct_ids()
    {
        var plan = StreamPlanner.Plan(
            Item(Fmt("image/png", MiB), Fmt("image/tiff", MiB)), peerStreams: true, NextId);
        Assert.Equal(2, plan.Streams.Count);
        Assert.NotEqual(plan.Streams[0].StreamId, plan.Streams[1].StreamId);
    }

    [Fact]
    public void Cap_drops_later_formats_in_order_and_keeps_earlier_ones()
    {
        // 60 + 50 > 100 MiB: the second is dropped, the third (small) still fits.
        var a = Fmt("image/png", 60 * MiB);
        var b = Fmt("image/tiff", 50 * MiB);
        var c = Fmt("text/plain", 100);
        var plan = StreamPlanner.Plan(Item(a, b, c), peerStreams: true, NextId);

        Assert.Equal(new[] { "image/png", "text/plain" }, plan.WireItem!.Formats.Select(f => f.Mime));
        var d = Assert.Single(plan.Dropped);
        Assert.Equal("image/tiff", d.Mime);
        Assert.Contains("100 MiB", d.Reason);
    }

    [Fact]
    public void Item_at_exactly_the_cap_is_kept()
    {
        var plan = StreamPlanner.Plan(Item(Fmt("image/png", 100 * MiB)), peerStreams: true, NextId);
        Assert.NotNull(plan.WireItem);
        Assert.Empty(plan.Dropped);
    }

    [Fact]
    public void Nothing_fits_yields_no_wire_item()
    {
        var plan = StreamPlanner.Plan(Item(Fmt("image/png", 100 * MiB + 1)), peerStreams: true, NextId);
        Assert.Null(plan.WireItem);
        Assert.Empty(plan.Streams);
        Assert.Single(plan.Dropped);
    }

    [Fact]
    public void Peer_without_stream_cap_gets_only_inline_sized_formats()
    {
        var text = Fmt("text/plain", 10);
        var img = Fmt("image/png", MiB);
        var plan = StreamPlanner.Plan(Item(text, img), peerStreams: false, NextId);

        Assert.Equal(new[] { "text/plain" }, plan.WireItem!.Formats.Select(f => f.Mime));
        Assert.Empty(plan.Streams);
        var d = Assert.Single(plan.Dropped);
        Assert.Contains("stream", d.Reason);
    }

    [Fact]
    public void Peer_without_stream_cap_and_only_large_formats_yields_nothing()
    {
        var plan = StreamPlanner.Plan(Item(Fmt("image/png", MiB)), peerStreams: false, NextId);
        Assert.Null(plan.WireItem);
    }

    [Fact]
    public void Wire_item_keeps_seq_origin_ts_and_hint()
    {
        var item = new ClipboardItem(42, Enumerable.Repeat((byte)7, 32).ToArray(), 999,
            new List<ClipFormat> { Fmt("image/png", MiB) }, "hint");
        var wire = StreamPlanner.Plan(item, peerStreams: true, NextId).WireItem!;
        Assert.Equal(42UL, wire.Seq);
        Assert.Equal(item.OriginDid, wire.OriginDid);
        Assert.Equal(999UL, wire.TsMs);
        Assert.Equal("hint", wire.Hint);
    }
}

File: clipsync-mac/Tests/ClipSyncTests/StreamPlannerTests.swift
import XCTest
@testable import ClipSync

final class StreamPlannerTests: XCTestCase {
    private let kib = 1024
    private let mib = 1024 * 1024
    private var next: UInt64 = 100
    private func nextId() -> UInt64 { defer { next += 1 }; return next }

    private func fmt(_ mime: String, _ size: Int) -> ClipFormat {
        ClipFormat(mime: mime, size: UInt64(size), payload: .inline(Data(count: size)))
    }
    private func item(_ formats: ClipFormat...) -> ClipboardItem {
        ClipboardItem(seq: 1, originDid: Data(count: 32), tsMs: 0, formats: formats, hint: nil)
    }
    private func isInline(_ f: ClipFormat) -> Bool { if case .inline = f.payload { return true } else { return false } }
    private func streamId(_ f: ClipFormat) -> UInt64? { if case .stream(let id) = f.payload { return id } else { return nil } }

    func testSmallItemPassesThroughInlineWithNoStreams() {
        let plan = StreamPlanner.plan(item(fmt("text/plain", 10), fmt("text/html", 64 * kib)),
                                      peerStreams: true, nextStreamId: nextId)
        XCTAssertNotNil(plan.wireItem)
        XCTAssertTrue(plan.wireItem!.formats.allSatisfy(isInline))
        XCTAssertTrue(plan.streams.isEmpty)
        XCTAssertTrue(plan.dropped.isEmpty)
    }

    func testLargeFormatIsReplacedByStreamIdAndQueued() {
        let big = fmt("image/png", 64 * kib + 1)
        let plan = StreamPlanner.plan(item(fmt("text/plain", 10), big), peerStreams: true, nextStreamId: nextId)
        let wire = plan.wireItem!
        XCTAssertEqual(wire.formats.count, 2)
        XCTAssertTrue(isInline(wire.formats[0]))
        XCTAssertNotNil(streamId(wire.formats[1]))
        XCTAssertEqual(wire.formats[1].size, big.size)
        XCTAssertEqual(plan.streams.count, 1)
        XCTAssertEqual(plan.streams[0].streamId, streamId(wire.formats[1]))
        XCTAssertEqual(plan.streams[0].data.count, 64 * kib + 1)
        XCTAssertTrue(plan.dropped.isEmpty)
    }

    func testDistinctStreamsGetDistinctIds() {
        let plan = StreamPlanner.plan(item(fmt("image/png", mib), fmt("image/tiff", mib)),
                                      peerStreams: true, nextStreamId: nextId)
        XCTAssertEqual(plan.streams.count, 2)
        XCTAssertNotEqual(plan.streams[0].streamId, plan.streams[1].streamId)
    }

    func testCapDropsLaterFormatsInOrderAndKeepsEarlierOnes() {
        let plan = StreamPlanner.plan(
            item(fmt("image/png", 60 * mib), fmt("image/tiff", 50 * mib), fmt("text/plain", 100)),
            peerStreams: true, nextStreamId: nextId)
        XCTAssertEqual(plan.wireItem!.formats.map(\.mime), ["image/png", "text/plain"])
        XCTAssertEqual(plan.dropped.count, 1)
        XCTAssertEqual(plan.dropped[0].mime, "image/tiff")
        XCTAssertTrue(plan.dropped[0].reason.contains("100 MiB"))
    }

    func testItemAtExactlyTheCapIsKept() {
        let plan = StreamPlanner.plan(item(fmt("image/png", 100 * mib)), peerStreams: true, nextStreamId: nextId)
        XCTAssertNotNil(plan.wireItem)
        XCTAssertTrue(plan.dropped.isEmpty)
    }

    func testNothingFitsYieldsNoWireItem() {
        let plan = StreamPlanner.plan(item(fmt("image/png", 100 * mib + 1)), peerStreams: true, nextStreamId: nextId)
        XCTAssertNil(plan.wireItem)
        XCTAssertTrue(plan.streams.isEmpty)
        XCTAssertEqual(plan.dropped.count, 1)
    }

    func testPeerWithoutStreamCapGetsOnlyInlineSizedFormats() {
        let plan = StreamPlanner.plan(item(fmt("text/plain", 10), fmt("image/png", mib)),
                                      peerStreams: false, nextStreamId: nextId)
        XCTAssertEqual(plan.wireItem!.formats.map(\.mime), ["text/plain"])
        XCTAssertTrue(plan.streams.isEmpty)
        XCTAssertEqual(plan.dropped.count, 1)
        XCTAssertTrue(plan.dropped[0].reason.contains("stream"))
    }

    func testPeerWithoutStreamCapAndOnlyLargeFormatsYieldsNothing() {
        let plan = StreamPlanner.plan(item(fmt("image/png", mib)), peerStreams: false, nextStreamId: nextId)
        XCTAssertNil(plan.wireItem)
    }

    func testWireItemKeepsSeqOriginTsAndHint() {
        let did = Data(repeating: 7, count: 32)
        let src = ClipboardItem(seq: 42, originDid: did, tsMs: 999, formats: [fmt("image/png", mib)], hint: "hint")
        let wire = StreamPlanner.plan(src, peerStreams: true, nextStreamId: nextId).wireItem!
        XCTAssertEqual(wire.seq, 42)
        XCTAssertEqual(wire.originDid, did)
        XCTAssertEqual(wire.tsMs, 999)
        XCTAssertEqual(wire.hint, "hint")
    }
}
